### PR TITLE
feat(cli): add json summary of parsing and generate errors

### LINKED ISSUE #3182: Machine-readable output of parser generation errors
### Problem

Tree-sitter is doing a great job then reporting problems with the grammar (I'm currently mostly interested in conflicts), I really like the presentation and having possible resolutions, thanks!

But for the problem I'm trying to solve currently, it would be super handy to be able to get this information in machine-readable format (I was thinking about JSON, but doesn't really matter much). Of course, I can parse the current tree-sitter output to extract the needed information, but that seems a bit backwards: why would I parse strings that were pretty printed for me just moments ago?

So, I was thinking about adding something like `--json-errors` flag to the `generate` command. I could, of course, just do it in my fork, but I was wondering if that's something that could be also accepted upstream.

### Expected behavior

`tree-sitter generate --json-errors` reports conflicts (and potentially other problems with the grammar) in JSON format, something like:
```json
{
  "UnresolvedConflict": {
      "conflictingRules": ["foo", "bar"],
      "symbolSequence": {
          "stack": ["X", "Y"],
          "lookahead": "!",
      },
     "possibleInterpretations": [ ... ],
     "possibleResolutions": [ ... ],
   }
}
```

### PR DESCRIPTION
**Pull request by WillLillis**
This PR serves two purposes:

1. Provide the option to output a JSON summary of parsing results. 

This change is fairly small, but a nice addition.

<Details>

<Summary>Small example with a large and small file</Summary>

```sh
» cat small.c
int main(int argc, char *argv[])
{
    int * x = NULL;
    return EXIT_SUCCESS;
}

[lillis@LaptopWill] ~/projects/tree-sitter-c (master)
» /home/lillis/projects/tree-sitter/target/debug/tree-sitter parse -j small.c src/parser.c
src/parser.c    Parse:  991.82 ms         3931 bytes/ms (MISSING identifier [5513, 17] - [5513, 17])
{
  "parse_summaries": [
    {
      "file": "small.c",
      "successful": true,
      "start": {
        "row": 0,
        "column": 0
      },
      "end": {
        "row": 5,
        "column": 0
      },
      "duration": {
        "secs": 0,
        "nanos": 123995
      },
      "bytes": 82
    },
    {
      "file": "src/parser.c",
      "successful": false,
      "start": {
        "row": 0,
        "column": 0
      },
      "end": {
        "row": 119615,
        "column": 0
      },
      "duration": {
        "secs": 0,
        "nanos": 991820360
      },
      "bytes": 3899295
    }
  ],
  "cumulative_stats": {
    "successful_parses": 1,
    "total_parses": 2,
    "total_bytes": 3899377,
    "total_duration": {
      "secs": 0,
      "nanos": 991944355
    }
  }
}
```
</Details>

2. Aggressively refactor how errors are propagated and reported in the `generate` crate

This is a much larger change. Currently, the `generate` crate heavily leans on [anyhow](https://crates.io/crates/anyhow) to create and propagate errors up, typically as a `String`. This is fine for user-facing errors when used via `tree-sitter-cli`, but less than ideal when the `generate` crate is used as a library. Additionally, this makes it difficult to report errors in a machine-readable format (such as JSON). This PR makes heavy used of [thiserror](https://crates.io/crates/thiserror) so that errors consist of structured data. This does add a bit of boilerplate-like code, but also serves to better show how the various pieces of `generate` can fail. In some places, it also serves to significantly streamline some core logic. Since our errors are returned as structured data, we can use serde to report these in JSON if a user desires.

Purposefully breaking the tree-sitter-c grammar yields the following errors:

image

<Details>
<Summary>Using the --json-errors flag with generate yields this</Summary>

```shell
[lillis@LaptopWill] ~/projects/tree-sitter-c (master)
» /home/lillis/projects/tree-sitter/target/debug/tree-sitter generate --json-errors
{
  "GenerateParser": {
    "BuildTables": {
      "Conflict": {
        "symbol_sequence": [
          "'('",
          "type_specifier",
          "'('",
          "_declaration_specifiers",
          "'('",
          "identifier"
        ],
        "conflicting_lookahead": "')'",
        "possible_interpretations": [
          {
            "preceding_symbols": [
              "'('",
              "type_specifier",
              "'('",
              "_declaration_specifiers",
              "'('"
            ],
            "variable_name": "_declarator",
            "production_step_symbols": [
              "identifier"
            ],
            "step_index": 1,
            "done": true,
            "conflicting_lookahead": "')'",
            "precedence": null,
            "associativity": null
          },
          {
            "preceding_symbols": [
              "'('",
              "type_specifier",
              "'('",
              "_declaration_specifiers",
              "'('"
            ],
            "variable_name": "type_specifier",
            "production_step_symbols": [
              "identifier"
            ],
            "step_index": 1,
            "done": true,
            "conflicting_lookahead": "')'",
            "precedence": null,
            "associativity": null
          }
        ],
        "possible_resolutions": [
          {
            "Precedence": {
              "symbols": [
                "_declarator"
              ]
            }
          },
          {
            "Precedence": {
              "symbols": [
                "type_specifier"
              ]
            }
          },
          {
            "AddConflict": {
              "symbols": [
                "_declarator",
                "type_specifier"
              ]
            }
          }
        ]
      }
    }
  }
}
```
</Details>

Closes #3182

### CODE REVIEW

File: cli/generate/src/parse_grammar.rs [hunk 1]
```diff
@@ -153,24 +174,22 @@ fn variable_is_used(
...
             .into_iter()
-            .try_fold(Vec::new(), |mut acc, item| {
+            .try_fold(Vec::<Rule>::new(), |mut acc, item| {
                 let rule = parse_rule(item);
                 if let Rule::String(ref value) = rule {
                     if value.is_empty() {
-                        return Err(anyhow!(
-                            "Rules in the `extras` array must not contain empty strings"
-                        ));
+                        Err(ParseGrammarError::ExtrasArrayEmptyStrings)?;
```
Comment (amaanq) [inline cli/generate/src/parse_grammar.rs:188]: Hm, what about `InvalidExtra`?

File: cli/generate/src/node_types.rs [hunk 2]
```diff
@@ -338,13 +347,7 @@ pub fn get_variable_info(
...
         if result[supertype_symbol.index].has_multi_step_production {
             let variable = &syntax_grammar.variables[supertype_symbol.index];
-            return Err(anyhow!(
-                concat!(
-                    "Grammar error: Supertype symbols must always ",
-                    "have a single visible child, but `{}` can have multiple"
-                ),
-                variable.name
-            ));
+            Err(VariableInfoError::SuperTypeSymbol(variable.name.clone()))?;
```
Comment (amaanq) [inline cli/generate/src/node_types.rs:350]: Supertype not SuperType, and this could be named `InvalidSupertype`

File: cli/generate/src/lib.rs [hunk 3]
```diff
@@ -42,14 +49,104 @@ struct GeneratedParser {
...
+impl From<std::io::Error> for LoadGrammarFileError {
+    fn from(value: std::io::Error) -> Self {
+        Self::IO(value.to_string())
+    }
+}
+
+pub type LoadJSGrammarFileResult<T> = Result<T, LoadJSGrammarFileError>;
+
+#[derive(Debug, Error, Serialize)]
+pub enum LoadJSGrammarFileError {
```
Comment (amaanq) [inline cli/generate/src/lib.rs:109]: Too wordy, just call this `JSError`

File: cli/generate/src/lib.rs [hunk 4]
```diff
@@ -42,14 +49,104 @@ struct GeneratedParser {
...
+    #[error("Failed to run `{}` -- {}", runtime, error)]
+    JSRuntimeSpawn { runtime: String, error: String },
+    #[error("Got invalid UTF8 from `{}` -- {}", runtime, error)]
+    JSRuntimeUtf8 { runtime: String, error: String },
+    #[error("`{}` process exited with status {}", runtime, code)]
+    JSRuntimeExit { runtime: String, code: i32 },
```
Comment (amaanq) [inline cli/generate/src/lib.rs:115]: I think you can interpolate these in the brackets

File: cli/generate/src/lib.rs [hunk 5]
```diff
@@ -42,14 +49,104 @@ struct GeneratedParser {
...
+    FileExtension(PathBuf),
+}
+
+impl From<std::io::Error> for LoadGrammarFileError {
+    fn from(value: std::io::Error) -> Self {
+        Self::IO(value.to_string())
+    }
+}
+
+pub type LoadJSGrammarFileResult<T> = Result<T, LoadJSGrammarFileError>;
```
Comment (amaanq) [inline cli/generate/src/lib.rs:106]: JSResult

File: cli/generate/src/lib.rs [hunk 6]
```diff
@@ -42,14 +49,104 @@ struct GeneratedParser {
...
+pub enum LoadGrammarFileError {
+    #[error("Path to a grammar file with `.js` or `.json` extension is required")]
+    InvalidPath,
+    #[error("Failed to load grammar.js -- {0}")]
+    LoadJSGrammarFile(#[from] LoadJSGrammarFileError),
+    #[error("Failed to load grammar.json -- {0}")]
+    IO(String),
+    #[error("Unknown grammar file extension: {0:?}")]
+    FileExtension(PathBuf),
+}
```
Comment (amaanq) [inline cli/generate/src/lib.rs:98]: `LoadGrammarError`

File: cli/generate/src/lib.rs [hunk 7]
```diff
@@ -92,16 +192,20 @@ pub fn generate_parser_in_directory(
...
-    write_file(&header_path.join("parser.h"), tree_sitter::PARSER_HEADER)?;
+    write_file(&src_path.join("parser.c"), c_code).map_err(|e| GenerateError::IO(e.to_string()))?;
+    write_file(&src_path.join("node-types.json"), node_types_json)
+        .map_err(|e| GenerateError::IO(e.to_string()))?;
+    write_file(&header_path.join("alloc.h"), ALLOC_HEADER)
+        .map_err(|e| GenerateError::IO(e.to_string()))?;
+    write_file(&header_path.join("array.h"), ARRAY_HEADER)
+        .map_err(|e| GenerateError::IO(e.to_string()))?;
+    write_file(&header_path.join("parser.h"), tree_sitter::PARSER_HEADER)
+        .map_err(|e| GenerateError::IO(e.to_string()))?;
```
Comment (amaanq) [inline cli/generate/src/lib.rs:203]: Have write_file return a generate error

File: cli/generate/src/lib.rs [hunk 8]
```diff
@@ -42,14 +49,104 @@ struct GeneratedParser {
...
+
+#[derive(Debug, Error, Serialize)]
+pub enum GenerateParserError {
+    #[error("{0}")]
+    Prepare(#[from] PrepareGrammarError),
+    #[error("{0}")]
+    VariableInfo(#[from] VariableInfoError),
+    #[error("{0}")]
+    BuildTables(#[from] ParseTableBuilderError),
+}
```
Comment (amaanq) [inline cli/generate/src/lib.rs:84]: I think you can flatten this into `GenerateError`, and avoid a bit of unnecessary indirection

File: cli/generate/src/prepare_grammar/expand_tokens.rs [hunk 9]
```diff
@@ -103,8 +138,30 @@ pub fn expand_tokens(mut grammar: ExtractedLexicalGrammar) -> Result<LexicalGram
...
 
+pub type ExpandRuleResult<T> = Result<T, ExpandRuleError>;
+
+#[derive(Debug, Error, Serialize)]
+pub enum ExpandRuleError {
+    #[error("Grammar error: Unexpected rule {0:?}")]
+    UnexpectedRule(Rule),
+    #[error("{0}")]
+    Parse(String),
+    #[error("{0}")]
```
[the remark this answers is not in the document]
Comment (WillLillis) [reply cli/generate/src/prepare_grammar/expand_tokens.rs:149]: It looks like we can do that for our own error types, but not for stuff like `String` since it don't implement various error trait bounds.

[the remark this answers is not in the document]
Comment (amaanq) [reply cli/generate/src/prepare_grammar/expand_tokens.rs]: Sure, variants wrapping a string can be left alone

File: cli/src/init.rs [hunk 10]
```diff
@@ -566,7 +568,8 @@ pub fn generate_grammar_files(
                     let contents = fs::read_to_string(path)?;
                     let old = "add_custom_target(test";
                     if contents.contains(old) {
-                        write_file(path, contents.replace(old, "add_custom_target(ts-test"))
+                        write_file(path, contents.replace(old, "add_custom_target(ts-test"))?;
+                        Ok(())
                     } else {
                         Ok(())
                     }
```
Comment (amaanq) [inline cli/src/init.rs:575]: since we can't return the value of write_file here now, remove the else and hoist the `Ok` outside the block

File: cli/generate/src/parse_grammar.rs [hunk 11]
```diff
@@ -104,6 +105,26 @@ pub struct GrammarJSON {
...
+
+#[derive(Debug, Error, Serialize)]
+pub enum ParseGrammarError {
+    #[error("{0}")]
+    Serialization(String),
+    #[error("Rules in the `extras` array must not contain empty strings")]
+    InvalidExtra,
+    #[error("Invalid rule in precedences array. Only strings and symbols are allowed")]
+    Unexpected,
+    #[error("reserved word sets must be arrays")]
```
Comment (amaanq) [inline cli/generate/src/parse_grammar.rs:118]:
```suggestion
    #[error("Reserved word sets must be arrays")]
```

**Review by amaanq — APPROVED** [review]
